### DISCUSSION TITLE
Connection close and stream data

### DISCUSSION THREAD

**Discussion by ManickaP** [microsoft/msquic#3804] [General]
Few questions on MsQuic internal behavior came up during our .NET discussion and we were wondering if you could shed light on them:

- What happens with stream data that are buffered, but not yet indicated to the user when the connection gets closed?
- Will they get discarded, because connection closure supersedes any stream events? 
- Is there a difference based on which side closed the connection?
- Is there a difference based on which data were ACKed and which weren't?
- Does received FIN flag play any role in it?

I'm aware that the connection should be given enough time to process all streams (e.g. idle out). But we are discussing corner scenarios and we want to make sure the .NET layer is not doing anything out of expectations for QUIC layer.

cc @wfurt @carnaviire

**Comment by nibanks** [comment]
Please clarify what you mean by "connection gets closed". Do you mean aborted/reset? Or do you mean gracefully shutdown? Aborted will abandon anything buffered. Graceful will not.

**Reply by nibanks** [reply]
Ah, nvm. I was thinking you meant the stream. If you kill a connection that is equivalent to an abortive shutdown of all streams.

**Reply by nibanks** [reply]
Now, that being said, as an implementation details, order matters. What I mean by that is the order you frame data into a QUIC packet, and the order you process it. In MsQuic, we ALWAYS put ACKs first in a packet, even before the CONNECTION_CLOSE frame. This means we do acknowledge and process completions as a result (i.e. send completion) before the connection shutdown. But this is not standard. Other implementations don't necessarily send any ACKs when they abort a connection. Other implementations might just ignore them on the receive side.

**Reply by ManickaP** [reply]
All the data gets discarded immediately and stream gets shutdown_complete, right?
Regardless whether we've received and ACKed all of them, including FIN flag, and potentially could still give them out to the user, right?

**Reply by nibanks** [reply]
Yes, abortive shutdown of a stream discards everything else queued for either send or receive.

**Reply by ManickaP** [reply]
Hmm, does this apply to SEND_COMPLETE? I.e. can we have `StreamSend` that never received its `SEND_COMPLETE` because of connection closure?

**Reply by nibanks** [reply]
You will get a send complete with the Canceled flag set.

**Reply by ManickaP** [reply]
Thank you!

**Reply by nibanks** [reply]
You're welcome!

**Reply by CarnaViire** [reply]
> Yes, abortive shutdown of a stream discards everything else queued for either send or receive.

But if we received FIN flag, doesn't it put stream's receive side into successful completion state? would that state be overridden?

What I mean is: writing side is still open and on reading side, the data, including FIN flag, is received by MsQuic but not fully handed down to user yet (e.g., callback for RECEIVE event returned Pending). Then peer tears down the connection. What will happen in that case?

**Reply by CarnaViire** [reply]
Also, from what I see in https://www.rfc-editor.org/rfc/rfc9000.html#section-3.2-13, it is up to the implementation to decide whether to discard the buffered data+FIN in case the stream is reset (or connection is closed), or to keep it and hand down to the user. Is that correct?

**Comment by Streamlet** [comment]
1. Call StreamSend
2. (then immediately) Call ConnectionShutdown with flag QUIC_CONNECTION_SHUTDOWN_FLAG_NONE

Is this a graceful shutdown or not?
Will the pending data in queue be sent before connection closed?

**Reply by nibanks** [reply]
Shutting down a connection always aborts all streams immediately.

**Reply by Streamlet** [reply]
How to wait streams to send all pending data and then close connection?

**Reply by nibanks** [reply]
Option 1: Use an idle timeout. This is what the HTTP protocol does. QUIC will naturally close the connection after no data is transferred for a period of time.

Option 2: Implement your own ref count based solution. When the last stream shuts down, terminate the connection immediately.
